sklearn/tests/test_init.py

```diff
@@ -18,3 +25,38 @@ def test_import_skl():
     # "import *" is discouraged outside of the module level, hence we
     # rely on setting up the variable above
     assert_equal(_top_import_error, None)
+
+
+def test_import_sklearn_no_warnings():
+    # Test that importing scikit-learn main modules doesn't raise any warnings.
+
+    try:
+        pkgs = pkgutil.iter_modules(path=sklearn.__path__, prefix='sklearn.')
+        import_modules = '; '.join(['import ' + modname
+                                    for _, modname, _ in pkgs
+                                    if (not modname.startswith('_') and
+                                        # add deprecated top level modules
+                                        # below to ignore them
+                                        modname not in [])])
+
+        message = subprocess.check_output(['python', '-Wdefault',
+                                           '-c', import_modules],
+                                          stderr=subprocess.STDOUT)
+        message = message.decode("utf-8")
+        message = '\n'.join([line for line in message.splitlines()
+                             if not (  # ignore ImportWarning
+                                     "ImportWarning" in line or
```

Comment:
In my case, I get,
```
$ python -Wdefault -c "import sklearn"                                                        [collections-abc-py37] 
/home/rth/.miniconda3/envs/sklearn-dev/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/home/rth/.miniconda3/envs/sklearn-dev/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
```
so this ignores these ImportWarnings

Reply:
do you know what that means? And does that mean we want to ignore all import warnings?

Reply:
There is a related discussion in https://github.com/astropy/astropy/issues/6025 suggesting some of it could come from Cython. In my case, it seems to be due to importing a pyx file in  MKL,
```py
PYTHONWARNINGS='error::ImportWarning' python -c "import sklearn"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/rth/src/scikit-learn/sklearn/__init__.py", line 64, in <module>
    from .base import clone
  File "/home/rth/src/scikit-learn/sklearn/base.py", line 10, in <module>
    import numpy as np
  File "/home/rth/.miniconda3/envs/sklearn-dev/lib/python3.6/site-packages/numpy/__init__.py", line 180, in <module>
    from . import fft
  File "/home/rth/.miniconda3/envs/sklearn-dev/lib/python3.6/site-packages/numpy/fft/__init__.py", line 14, in <module>
    import mkl_fft._numpy_fft as _nfft
  File "/home/rth/.miniconda3/envs/sklearn-dev/lib/python3.6/site-packages/mkl_fft/__init__.py", line 27, in <module>
    from ._pydfti import (fft, ifft, fft2, ifft2, fftn, ifftn, rfft, irfft,
  File "mkl_fft/_pydfti.pyx", line 27, in init mkl_fft._pydfti
ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
```

in any case it seems to be upstream issues, that will not be seen by users (as `ImportWarning` is ignored by default), I'm not sure it's worth the effort investigating further.

Reply:
fair, maybe add a comment to this effect?